### ISSUE TITLE
Problems with intent_filter

### ISSUE THREAD

**Issue by maxbachmann** [rhasspy/rhasspy-rasa-nlu-hermes#4]
As far as I can see in the code the intent_filter simply get's ignored for the rasa nlu

**Comment by maxbachmann** [comment]
https://github.com/rhasspy/rhasspy-rasa-nlu-hermes/pull/1 uses the intent filter, but since it simply filters the intents after detection we will still run into issues when e.g. a user has to apps that use a `yes/no` intent, since there is no way for the NLU to know which one the user means without providing it with the intent_filter. Since apparently the rasa nlu does not have to capability we might want to use a fallback NLU when a intent is detected, but not part of the `intent_filter`

**Comment by daniele-athome** [comment]
I'm sorry I don't understand. If the app wants to know either "yes" or "no" it could provide an intent filter with 2 intents, e.g. `["YesIntent", "NoIntent"]`. Right?

**Comment by maxbachmann** [comment]
We could have `app A` and `app B` with the intents `["A:YesIntent", "A:NoIntent"]` and `["B:YesIntent", "B:NoIntent"]`, since they both require user interaction. So in snips as you described here: https://community.rhasspy.org/t/only-allow-some-intents-to-start-a-dialogue/973/4 you would deactivate them by default and then when you need them in the conversation activate them. However the way the intent filter is implemented here you might pass `intent_filter=["A:YesIntent", "A:NoIntent"]`, but rasa detects `B:YesIntent`, since they basically have the same training examples. Then afterwards we would detect that `B:YesIntent`, is not in `["A:YesIntent", "A:NoIntent"]` -> intent not detected even though it should be able to when passing it this intent filter.

**Comment by daniele-athome** [comment]
Oh I got it now. I didn't think of that. Damn, that's a serious problem. I could ask over the guys at Rasa though, so it may be implemented by the HTTP API.
A "dirty" workaround would be to use common intents, but [it has other pitfalls](https://community.rhasspy.org/t/helper-library-to-develop-rhasspy-apps-in-python/969/5?u=daniele_athome).

**Comment by maxbachmann** [comment]
I think asking the guys at rasa would make sence, since any solution that does not involve the nlu directly will always be a workaround (or they might already know people that had similar problems and how they solved it).

**Comment by daniele-athome** [comment]
In the meantime... [Rasa does give intent rankings](https://rasa.com/docs/rasa/api/http-api/#operation/parseModelMessage)... we could use those maybe (as workaround): fallback to looking into intent rankings when the main intent is filtered. The problem is that intent rankings don't have entities. I know, very ugly workaround :-)

**Comment by koenvervloesem** [comment]
I still think using common intents for things like "Yes", "No" is more reasonable. What's the problem with this approach? I don't see any pitfall for this situation in the forum post you linked to, @daniele-athome. As long as there's only one session active at a time on each site ID, so only one app at a time filters and reacts on these intents, this is fine, isn't it?

**Comment by maxbachmann** [comment]
How would you implement this? I do not think we can add all the intents that might reappear as build in (there are simply to many) and I do not thinks it works to tell independent skill developers that they need to name their intents similar to everyone else creating skills. (the intents might not even be 100% the same, but only share part of their training data)

**Comment by koenvervloesem** [comment]
No, not for all of them, I don't think we should define a whole common intent collection. But "Yes" and "No" seems like something that many apps would need. If you let app developers define the example sentences for these intents themselves, you end up with one app that detects "Yeah" as intent Yes, and another one that doesn't, just because the developer of the former app has thought about this example sentence and the developer of the latter hasn't. This will lead to a lot of confusion for users.

And sure, some apps will probably want to create variations on a simple "Yes" or "No", but they can add a separate intent YesWithExtras with extra example sentences and let their app react to both the default Yes and their own YesWithExtras. And of course every developer is free to use "built-in" intents or not.

I just think users will benefit from consistency of the voice interface, and every app developer having to come up with examples for basic intents like "Yes" and "No" will result in a wildly inconsistent voice interface. If there are some technical hurdles (are there?) I think we should try to overcome them. Rhasspy is still young, now is the moment to think these issues through to be able to create a system that is as user-friendly and developer-friendly as possible.

**Comment by maxbachmann** [comment]
Yes I agree that a build in intent/slot for yes/no could make sence.
I just thought you ment this to solve the problem of this issue, since it does fix it only for some of the cases ;)
E.g. in rhasspy-fuzzywuzzy this problem does not exist, since it filters the choices available for the nlu using the intent_filter before it even matches them.

**Comment by koenvervloesem** [comment]
Yes, I should have made it clear that I was reacting on the Yes/No examples specifically. This doesn't solve the issue you raised.

**Comment by daniele-athome** [comment]
[I've created a topic on Rasa's forum](https://forum.rasa.com/t/restrict-nlu-parsing-to-a-subset-of-intents/28905/2), feel to correct or integrate it in any way you like.

Anyway, I feel like I don't understand Rasa completely. I mean Rhasspy only uses the NLU part, but the more I study Rasa features the more I think that this might be achieved with other Rasa features such as actions, stories, etc. I don't to want to tight Rhasspy too much to Rasa though.

**Comment by maxbachmann** [comment]
Thanks :+1: 

I am not really familiar with rasa either, but it appears that when you want to use mutliple different conversation paths, these need to be known when training already, which we do not in Rhasspy right now

**Comment by daniele-athome** [comment]
Ok I studied the issue a little more, but it requires more research. [The guys at Rasa's made me understand a few things about how Rasa actually works](https://forum.rasa.com/t/restrict-nlu-parsing-to-a-subset-of-intents/28905/5) and this might be more complicated than I anticipated.
